lib/datasource/artifactory/index.ts

```diff
@@ -0,0 +1,97 @@
+import { logger } from '../../logger';
+import { ExternalHostError } from '../../types/errors/external-host-error';
+import { cache } from '../../util/cache/package/decorator';
+import { parse } from '../../util/html';
+import { HttpError } from '../../util/http/types';
+import { Datasource } from '../datasource';
+import type { GetReleasesConfig, Release, ReleaseResult } from '../types';
+import { datasource, defaultRegistryUrl } from './common';
+
+export class ArtifactoryDatasource extends Datasource {
+  static readonly id = datasource;
+
+  constructor() {
+    super(datasource);
+  }
+
+  override readonly customRegistrySupport = true;
+
+  override readonly defaultRegistryUrls = [defaultRegistryUrl];
+
+  override readonly caching = true;
+
+  override readonly registryStrategy = 'merge';
+
+  @cache({
+    namespace: `datasource-${datasource}`,
+    key: ({ registryUrl }: GetReleasesConfig) => `${registryUrl}`,
+  })
+  async getReleases({
+    lookupName,
+    registryUrl,
+  }: GetReleasesConfig): Promise<ReleaseResult | null> {
+    if (registryUrl === defaultRegistryUrl) {
+      logger.warn(
+        'artifactory datasource requires custom registryUrl. Skipping datasource'
+      );
+      return null;
+    }
+
+    const url = `${registryUrl}/${lookupName}`;
+
+    const result: ReleaseResult = {
+      homepage: 'https://jfrog.com/artifactory',
+      releases: [],
+    };
+    try {
+      const response = await this.http.get(url);
+      const body = parse(ArtifactoryDatasource.cleanSimpleHtml(response.body));
+      const nodes = body.querySelectorAll('a');
+
+      let candidates: string[] = [];
+      nodes.forEach((node) => candidates.push(node.innerHTML));
+
+      // filter out hyperlink to navigate to parent folder
+      candidates = candidates.filter(
+        (candidate) => candidate !== '../' && candidate !== '..'
+      );
+
+      candidates.forEach((candidate) => {
+        const parsedCandidate: string =
+          candidate.slice(-1) === '/' ? candidate.slice(0, -1) : candidate;
+
+        const thisRelease: Release = {
+          version: '',
+        };
+        thisRelease.version = parsedCandidate;
+        result.releases.push(thisRelease);
+      });
+
+      if (result.releases.length) {
+        logger.debug('Found ' + String(result.releases.length) + ' releases');
```

Comment:
This doesn't give enough context. You'll just see messages like "Found 2 releases" which don't tell you the datasource or package name. Better to lower this to trace too

Reply:
Lowered to trace and added more context too.